### PR TITLE
Modal type bug fix grab bag

### PR DESCRIPTION
**Pull request by iwehrman**
This is modal-type bug-fix PR. I believe it addresses issues #2071, #2009, #1927, #1883, #1881, #1857, #1691. There are still a number of type issues it does not address, so please take a look at the [Type issue label](https://github.com/adobe-photoshop/spaces-design/labels/Type) if you see a problem during testing to see if an issue has already been filed.

Highlights:

* Removes the `*Assigned` properties in `ParagraphStyle` and `CharacterStyle` in favor of a simpler `mergeWith` operation for resolving inconsistencies.
* Simplifies and throttles the `updatedTextProperties` event handler, which PS emits whenever the style properties for the current cursor position/text selection change in the modal state.
* Applies this handler in the select type sub-tool, as well as in the full type tool.
* Makes use of a new `deleteTextLayer` event to catch the case of a new text layer being canceled without committing.
* Factors a `layers.removeLayers` action out of `deleteSelected`, which can be used to remove layers from the JS-side layers model.
* Adds a `type.updateProperties` action to set type properties en masse instead of with separate actions that redraw the UI multiple times redundantly. (TODO: eliminate the other `type.update*` actions in favor of this one.)

@mcilroyc: I'm not sure if I've used the history events correctly. Can you double-check that part of this PR for me?

Requires at least Mac Jenkins pgdev.469 from the afternoon of Monday the 17th.

### CODE REVIEW

File: src/js/models/characterstyle.js [hunk 1]
```diff
@@ -119,23 +96,17 @@ define(function (require, exports, module) {
             textStyle = characterStyleDescriptor.textStyle;
 
         var rawColor = textStyle.hasOwnProperty("color") ?
-            textStyle.color :
-            baseParentStyle.color,
-            color = Color.DEFAULT;
+            textStyle.color : baseParentStyle.color;
```
Comment (volfied) [inline src/js/models/characterstyle.js]: can you indent this, it's hard to follow.

File: src/js/models/characterstyle.js [hunk 2]
```diff
@@ -187,64 +140,67 @@ define(function (require, exports, module) {
...
-            }, new CharacterStyle());
+                return CharacterStyle.fromCharacterStyleDescriptor(documentDescriptor,
+                    layerDescriptor, characterStyleDescriptor, baseParentStyle);
+            });
+
+        var reducedModel = Object.keys(_characterStylePrototype)
+            .reduce(function (reducedModel, property) {
+                var values = collection.pluck(characterStyles, property);
+
+                reducedModel[property] = collection.uniformValue(values);
```
Comment (volfied) [inline src/js/models/characterstyle.js]: No protection against null?

File: src/js/tools/superselect/type.js [hunk 3]
```diff
@@ -43,6 +53,14 @@ define(function (require, exports, module) {
...
             }
         });
+
+        if (_typeChangedHandler) {
+            descriptor.removeListener("updateTextProperties", _typeChangedHandler);
+        }
+
+        _typeChangedHandler = TypeTool.updateTextPropertiesHandler.bind(this);
+
+        descriptor.addListener("updateTextProperties", _typeChangedHandler);
```
Comment (volfied) [inline src/js/tools/superselect/type.js]: should deselectHandler be removing this listener?

File: src/js/models/characterstyle.js (hunk 2 above)
Comment (iwehrman) [inline src/js/models/characterstyle.js]: In the case of these reduce models, null will be interpreted in the UI to mean "mixed".

File: src/js/tools/superselect/type.js (hunk 3 above)
Comment (iwehrman) [inline src/js/tools/superselect/type.js]: Mmm, yeah. Good call.

File: src/js/models/characterstyle.js (hunk 1 above)
Comment (iwehrman) [inline src/js/models/characterstyle.js]: OK, sure.

**Comment by iwehrman** [comment]
Fixes in commit 912c291.

**Comment by volfied** [comment]
Not merging this yet as I'd like to go through the bugs and code one more time.

File: src/js/actions/layers.js [hunk 4]
```diff
@@ -790,9 +790,40 @@ define(function (require, exports) {
...
+        return _getSelectedLayerIndices(document)
+            .bind(this)
+            .then(function (selectedLayerIndices) {
+                var payload = {
+                    documentID: document.id,
+                    layerIDs: collection.pluck(layers, "id"),
+                    selectedIndices: selectedLayerIndices
+                };
+
+                this.dispatch(events.document.history.nonOptimistic.DELETE_LAYERS, payload);
```
Comment (mcilroyc) [inline src/js/actions/layers.js]: It seems like this *might* need to call a non-history flavor of DELETE_LAYERS when called by the "deleteTextLayer" handler.  I need to get the newest build to test this - just wanted to register the suspicion.

Comment (iwehrman) [inline src/js/actions/layers.js]: Do you suggest a flag?

Comment (mcilroyc) [inline src/js/actions/layers.js]: I believe this scenario (clicking canvas to create a text layer and then ESCing before typing) causes two extraneous history states in our store that do not have corresponding PS history.  These are created by the layer actions `addLayers()` and `removeLayers()`, both of which dispatch history-changing events.

You can test this by creating a rectangle, changing its color twice, doing the point-text-empty-ESC dance, and then undoing twice.  You should see the rectangle change color on canvas, but not updated in the panel.

It seems like a simple fix for the removeLayers case (conditionally dispatching a non-history version of the event).  If it is not that simple for the addLayers case, and we therefore might need to do some retrospective history manipulation, then LMK.

**Comment by volfied** [comment]
#2081 is still not fixed in this PR.

**Comment by volfied** [comment]
#1882 is still not fixed.

**Comment by volfied** [comment]
Other issues listed at the top are good. At this point I'm passing this review to @mcilroyc so he can wrap up the history code review.

**Comment by iwehrman** [comment]
@volfied: #1927 seems fixed to me. Can you explain how to repro a problem? 
Re #1882: I may have misunderstood the bug originally, but I'm not sure PS even supports paragraph alignments at the text-run level?

**Comment by volfied** [comment]
@iwehrman #1927: Select a shape layer and a text layer (I selected shape layer last) and try changing the text color. Now I realize I may have to rebuild (I thought i rebuilt yesterday after 2:30 PM when Jeffrey checked his changes in).

#1882 - If paragraph alignment is not usable during modal type state, we should remove/disable those buttons.

**Comment by iwehrman** [comment]
@volfied: that's what I'm doing for #1927.
@pineapplespatula: should the alignment buttons be disabled during modal text editing?

**Comment by iwehrman** [comment]
It looks like the alignment buttons _do_ work in standard PS, so I'll see if either a) the action descriptors aren't working, in which case I think we should proceed and leave #1882 open, or b) if they are working, I'll fix it.

**Comment by pineapplespatula** [comment]
@iwehrman the alignment buttons should be enabled. I'm pretty sure this one is a core bug

**Comment by iwehrman** [comment]
The next check-in will also address #1883.

**Comment by iwehrman** [comment]
@pineapplespatula yes, that's consistent with what I'm seeing. I officially retract that this addresses #1882.

**Comment by iwehrman** [comment]
The next check-in will also address #1881.

**Comment by volfied** [comment]
#1927 confirmed, I had an earlier build :X

**Comment by volfied** [comment]
#1881 fix confirmed. @mcilroyc, please merge when ready.

**Comment by iwehrman** [comment]
I've pushed the _narrow_ fix for the history state problem discussed after the standup this morning. I notice that once I clear the document history, though, undo/redo are no longer enabled despite the fact that the document is still dirty. Is this a known issue? It seems independent of what I'm doing here. In any case, back to @mcilroyc for review.

**Comment by iwehrman** [comment]
Also note again that 7c6bfa4 claims to address #1881 and #1883, but I think @volfied only verified the former.

File: src/js/actions/layers.js [hunk 5]
```diff
@@ -790,9 +791,46 @@ define(function (require, exports) {
...
+
+                if (resetHistory) {
+                    return this.transfer(historyActions.resetHistory, document);
```
Comment (mcilroyc) [inline src/js/actions/layers.js]: Per the above-noted issue with undo/redo not being available in the menu, try...

changing this line to `return this.transfer(historyActions.queryCurrentHistory, document.id);`
(and the associated `.transfers` change)

I'm not saying this is the ideal solution, but it is more correct because it asks photoshop what the current pointer should be at, and then re-inits the  history store with its known-unknowns

This throws a warning ...which could probably be forcibly quelled with a `iKnowWhatImDoing` flag

File: src/js/tools/type.js [hunk 6]
```diff
@@ -70,80 +84,66 @@ define(function (require, exports, module) {
...
+                    return;
+                }
+
+                var layerID = event.layerID,
+                    layer = document.layers.byID(layerID);
+
+                if (layer) {
+                    if (_layersReplaced) {
+                        // See comment above at the _layersReplaced declaration
+                        this.flux.actions.documents.updateDocument();
```
Comment (mcilroyc) [inline src/js/tools/type.js]: should _layersReplaced be set to false here, so the subsequent deleteTextLayer events aren't misinterpreted, assuming there was *not* a createTextLayer in the interim?  Maybe that's not possible?

Comment (iwehrman) [inline src/js/tools/type.js]: Shouldn't be possible? But, better safe than sorry?

Comment (iwehrman) [inline src/js/tools/type.js]: Actually I think I'm against adding that extra line, since it will make the supposed state machine more obscure from our end, and there's no way to defend against all possible PS bugs.

**Comment by mcilroyc** [comment]
Alright, droppin this bomb.

**Comment by chadrolfs** [comment]
set appropriate labels and re-assigned bugs addressed in this PR.
